src/modules/CaaSMapper.ts

```diff
@@ -737,26 +827,50 @@ export class CaaSMapper {
               remoteProject: remoteProjectId,
               filterContext,
             },
-            this.referenceDepth
-          )
-        )
-      )
-      const fetchedItems = response.map(({ items }) => items).flat()
-      ids.forEach((id) =>
-        referencedItems[id].forEach((path) =>
-          set(
-            data,
-            path,
-            fetchedItems.find((data) => {
-              const hasId = typeof data === 'object' && 'id' in (data as object)
-              if (hasId) {
-                return (data as { id: string }).id === id
-              }
-            })
+            this
           )
         )
       )
     }
-    return data
+  }
+
+  // we use a query function, because ids get mixed up:
+  // referencedItems store identifier, resolvedReferences store _id or previewId
+  static findResolvedReferencesByIds(ids: string[], resolvedReferences: ResolvedReferencesInfo) {
+    const items = []
+    for (const [resolvedId, resolvedItem] of Object.entries(resolvedReferences)) {
+      for (const id of ids) {
+        if (id.includes(resolvedId) || resolvedId.includes(id)) {
+          items.push(resolvedItem)
+        }
+      }
+    }
+    return items
+  }
+
+  static denormalizeResolvedReferences(
+    mappedItems: (CaasApi_Item | MappedCaasItem)[],
+    referenceMap: ReferencedItemsInfo,
+    resolvedReferences: ResolvedReferencesInfo
+  ) {
+    if (!referenceMap || Object.keys(referenceMap).length === 0) return mappedItems
+    // denormalize
+    for (const [referencedId, occurencies] of Object.entries(referenceMap)) {
+      occurencies.forEach((path) => {
+        // no simple comparison possible because referencedItems store identifier
+        // and resolvedReferences store "previewId" or "_id"
+        const resolvedId = Object.keys(resolvedReferences).find((key) => key.includes(referencedId))
```

Comment:
If we cannot refactor that as a simple object access, at least Object.keys should be put out of the loop